patches/server/0925-Add-Offline-PDC-API.patch

```diff
@@ -0,0 +1,99 @@
++    }
++    // Paper end
+ 
+     @Nullable
+     public CompoundTag load(Player player) {
+diff --git a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
+index 6d2ba650f53de8a460857f1846401a20b50cc43c..527e602ed47d11442b57eefbf3566b7793aa36db 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
+@@ -304,6 +304,48 @@ public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializa
+ 
+         return result;
+     }
++
++    @Override
++    public org.bukkit.persistence.PersistentDataContainer getPersistentDataContainer() {
++        Player player = getPlayer();
++        if(player != null) return player.getPersistentDataContainer();
++
++        CompoundTag result = getData();
++        if(result == null) return null; // never called
++
++        org.bukkit.craftbukkit.persistence.CraftPersistentDataContainer pdc = new org.bukkit.craftbukkit.persistence.CraftPersistentDataContainer(org.bukkit.craftbukkit.entity.CraftEntity.DATA_TYPE_REGISTRY) {
++
++            private void save() {
++                if (!isEmpty()) {
++                    result.put("BukkitValues", toTagCompound());
++                } else {
++                    result.remove("BukkitValues");
++                }
++                storage.save(getUniqueId(), result);
++            }
++
++            @Override
++            public <T, Z> void set(org.bukkit.NamespacedKey key, org.bukkit.persistence.PersistentDataType<T, Z> type, Z value) {
++                super.set(key, type, value);
++                save();
++            }
++
++            @Override
++            public void remove(org.bukkit.NamespacedKey key) {
++                super.remove(key);
++                save();
++            }
++        };
++
++        if (result != null) {
++            net.minecraft.nbt.Tag base = result.get("BukkitValues");
++            if (base instanceof CompoundTag compound) {
++                pdc.putAll(compound);
++            }
++        }
++
++        return pdc;
++    }
+     // Paper end
+ 
+     @Override
+diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+index 2a6c67634c31c332102d24bef293da1bacd0c000..424a5a8a8345c89dda54687aaf74be3b69f81ca3 100644
+--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+@@ -190,7 +190,7 @@ import org.bukkit.util.Vector;
+ 
+ public abstract class CraftEntity implements org.bukkit.entity.Entity {
+     private static PermissibleBase perm;
+-    private static final CraftPersistentDataTypeRegistry DATA_TYPE_REGISTRY = new CraftPersistentDataTypeRegistry();
++    public static final CraftPersistentDataTypeRegistry DATA_TYPE_REGISTRY = new CraftPersistentDataTypeRegistry(); // Paper private -> public
```

Comment:
If you want, this could be in the access transformers instead of a manual change.
Up to you, I don't think we really have much of a preference.

Reply:
I'd leave it as a change here. I think we are trying to move away from making everything an AT, leaving that for larger things, or things that need to change visibility across inheritance.